File: 200-number-of-islands/number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        num_island = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    self.dfs(grid, i, j)
                    num_island += 1
        return num_island
    
    
    def dfs(self, grid, r, c):
        if(
            r < 0
            or c < 0
            or r >= len(grid)
            or c >= len(grid[0])
            or grid[r][c] != '1'
        ):
            return
        grid[r][c] = "0"

        self.dfs(grid, r - 1, c)
        self.dfs(grid, r + 1, c)
        self.dfs(grid, r, c - 1)
        self.dfs(grid, r, c + 1)

        

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int count = 0;
        for(int i  = 0; i<rows ; i++){
            for(int j = 0; j<cols; j++){
                if (grid[i][j] =='1'){
                    count++;
                    dfs(grid, i,j);

                }
            }
        }
        return count;

        

    }

    private void dfs(char[][] grid, int r, int c){
        if(r<0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] != '1' ){
            return;
        }
        grid[r][c] = '0';
        dfs(grid ,r-1,c);
        dfs(grid ,r+1,c);
        dfs(grid ,r,c-1);
        dfs(grid ,r,c+1);
    }


}